### ISSUE TITLE
kube-scheduler panic

### ISSUE THREAD

**Issue by jacobstr** [kubernetes/kubernetes#70450]
**What happened**: kube-scheduler crashes (panics) in rare circumstances. Stack trace:

```
Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:573
/usr/local/go/src/runtime/panic.go:502
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:388
/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/cache/node_info.go:386
/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:909
/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:831
/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:47
/usr/local/go/src/runtime/asm_amd64.s:2361
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x161223a]
goroutine 44674294 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x107
panic(0x418ba40, 0x85fd7f0)
	/usr/local/go/src/runtime/panic.go:502 +0x229
k8s.io/kubernetes/pkg/scheduler/cache.(*NodeInfo).Clone(0x0, 0xc428b44ac0)
	/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/cache/node_info.go:386 +0x3a
k8s.io/kubernetes/pkg/scheduler/core.selectVictimsOnNode(0xc42e66c700, 0x60fa9e0, 0xc422c60630, 0x0, 0xc4215a5e00, 0x61505a0, 0xc4201a7ab0, 0xc425744980, 0x5, 0x8, ...)
	/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:909 +0x7d
k8s.io/kubernetes/pkg/scheduler/core.selectNodesForPreemption.func1(0xf)
	/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:831 +0x176
k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.Parallelize.func1(0xc4300e3c50, 0xc4302c94a0, 0xc429e60380)
	/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:47 +0x96
created by k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.Parallelize
	/workspace/anago-v1.11.2-beta.0.50+bb9ffb1654d4a7/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:43 +0x10b
```

**What you expected to happen**: flawless operation ;)

**How to reproduce it (as minimally and precisely as possible)**: Unclear. The error above looks like it was associated with some cloning of node metadata. Perhaps due to a race during node scale-up/scale-down and cache interactions.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):

```
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.2", GitCommit:"bb9ffb1654d4a729bb4cec18ff088eacc153c239", GitTreeState:"clean", BuildDate:"2018-08-07T23:08:19Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
```

/kind bug

**Comment by Huang-Wei** [comment]
@jacobstr are you using hostPort for your workload?

**Comment by jacobstr** [comment]
@Huang-Wei is there a specific workload you had in mind? The `hyperkube apiserver` does, as well as  daemonsets for our service mesh, which are in host networking mode.

**Comment by Huang-Wei** [comment]
There was a PR (#67456) in v1.12/v1.11.3 fixing issue when workloads uses hostNetwork, and the solution is to correct the NodeInfo cloning logic. So it's quite likely related with the symptom you described.

Could you give a try on v1.11.3?

**Comment by Huang-Wei** [comment]
BTW: the issue that PR fixed is only reproducible on some circumstances.

**Comment by jacobstr** [comment]
Thanks @Huang-Wei - I'll give this a whirl. I think I can optimistically close this issue - we'll reopen if it persists with a more recent autoscaler release.

Cheers!

**Comment by patrickshan** [comment]
@Huang-Wei we just had a similar issue this morning and our cluster is running kube 1.11.4:

`Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.4", GitCommit:"bf9a868e8ea3d3a8fa53cbb22f566771b3f8068b", GitTreeState:"clean", BuildDate:"2018-10-25T19:06:30Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}`

and these are the error logs : 

`E1114 18:58:19.944359       1 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/workspace/anago-v1.11.4-beta.0.135+bf9a868e8ea3d3/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/workspace/anago-v1.11.4-beta.0.135+bf9a868e8ea3d3/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/workspace/anago-v1.11.4-beta.0.135+bf9a868e8ea3d3/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:573
/usr/local/go/src/runtime/panic.go:502
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:388
/workspace/anago-v1.11.4-beta.0.135+bf9a868e8ea3d3/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/cache/node_info.go:386
/workspace/anago-v1.11.4-beta.0.135+bf9a868e8ea3d3/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:910
/workspace/anago-v1.11.4-beta.0.135+bf9a868e8ea3d3/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:832
/workspace/anago-v1.11.4-beta.0.135+bf9a868e8ea3d3/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:47
/usr/local/go/src/runtime/asm_amd64.s:2361
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x1613aea]

goroutine 22484147 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/workspace/anago-v1.11.4-beta.0.135+bf9a868e8ea3d3/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x107
panic(0x4190cc0, 0x86077f0)
	/usr/local/go/src/runtime/panic.go:502 +0x229
k8s.io/kubernetes/pkg/scheduler/cache.(*NodeInfo).Clone(0x0, 0xc422217ba0)
	/workspace/anago-v1.11.4-beta.0.135+bf9a868e8ea3d3/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/cache/node_info.go:386 +0x3a
k8s.io/kubernetes/pkg/scheduler/core.selectVictimsOnNode(0xc43b718a80, 0x6101040, 0xc436333950, 0x0, 0xc42167a030, 0x6156c60, 0xc4206a5810, 0xc42da8e6f0, 0x2, 0x2, ...)
	/workspace/anago-v1.11.4-beta.0.135+bf9a868e8ea3d3/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:910 +0x7d
k8s.io/kubernetes/pkg/scheduler/core.selectNodesForPreemption.func1(0x6)
	/workspace/anago-v1.11.4-beta.0.135+bf9a868e8ea3d3/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:832 +0x176
k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.Parallelize.func1(0xc4300da800, 0xc42c40e000, 0xc42ed17980)
	/workspace/anago-v1.11.4-beta.0.135+bf9a868e8ea3d3/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:47 +0x96
created by k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.Parallelize
	/workspace/anago-v1.11.4-beta.0.135+bf9a868e8ea3d3/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:43 +0x10b
`
we do have hostNetwork work load running inside our cluster but not sure if that's caused by other issues since you mentioned that hostNetwork issue has been fixed since 1.11.3.

**Comment by patrickshan** [comment]
@jacobstr could you re-open this issue ? or I can raise a new issue for it.

**Comment by Huang-Wei** [comment]
/reopen

I will take a look.

**Comment by patrickshan** [comment]
@Huang-Wei thanks

**Comment by Huang-Wei** [comment]
@patrickshan are the worker nodes auto scaling out/in? e.g. by an auto-scaler component?

**Comment by patrickshan** [comment]
@Huang-Wei yes, we do have an autoscaler to auto scale our worker nodes and we are using https://github.com/atlassian/escalator since most of our workloads are job based.

**Comment by Huang-Wei** [comment]
I see. It's likely an internal cache inconsistency. (not related with the hostPort issue I fixed earlier)

/assign

**Comment by Huang-Wei** [comment]
One thing I'm curious is from the log it points to L386 of node_info.go:

https://github.com/kubernetes/kubernetes/blob/bf9a868e8ea3d3a8fa53cbb22f566771b3f8068b/pkg/scheduler/cache/node_info.go#L386

```
goroutine 22484147 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/workspace/anago-v1.11.4-beta.0.135+bf9a868e8ea3d3/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x107
panic(0x4190cc0, 0x86077f0)
	/usr/local/go/src/runtime/panic.go:502 +0x229
k8s.io/kubernetes/pkg/scheduler/cache.(*NodeInfo).Clone(0x0, 0xc422217ba0)
	/workspace/anago-v1.11.4-beta.0.135+bf9a868e8ea3d3/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/cache/node_info.go:386 +0x3a
k8s.io/kubernetes/pkg/scheduler/core.selectVictimsOnNode(0xc43b718a80, 0x6101040, 0xc436333950, 0x0, 0xc42167a030, 0x6156c60, 0xc4206a5810, 0xc42da8e6f0, 0x2, 0x2, ...)
	/workspace/anago-v1.11.4-beta.0.135+bf9a868e8ea3d3/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:910 +0x7d
k8s.io/kubernetes/pkg/scheduler/core.selectNodesForPreemption.func1(0x6)
	/workspace/anago-v1.11.4-beta.0.135+bf9a868e8ea3d3/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:832 +0x176
k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.Parallelize.func1(0xc4300da800, 0xc42c40e000, 0xc42ed17980)
	/workspace/anago-v1.11.4-beta.0.135+bf9a868e8ea3d3/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:47 +0x96
created by k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.Parallelize
	/workspace/anago-v1.11.4-beta.0.135+bf9a868e8ea3d3/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:43 +0x10b
```

Then it means `n.requestedResource` is nil, but should't `n.requestedResource.Clone()` give a panic trace at the first call inside `n.requestedResource.Clone()`? Say:

https://github.com/kubernetes/kubernetes/blob/bf9a868e8ea3d3a8fa53cbb22f566771b3f8068b/pkg/scheduler/cache/node_info.go#L194

I looked through the code, but can't find a place that `n.requestedResource` could be nil. So @patrickshan are you using raw v1.11.4-beta.0 branch, or did some modification? As in my opinion, it makes sense the panic comes from L385:

https://github.com/kubernetes/kubernetes/blob/bf9a868e8ea3d3a8fa53cbb22f566771b3f8068b/pkg/scheduler/cache/node_info.go#L385

**Comment by bsalamat** [comment]
> Then it means n.requestedResource is nil, but should't n.requestedResource.Clone() give a panic trace at the first call inside n.requestedResource.Clone()?

No, in Go if n.requestedResource is nil, you will get a stack trace similar to what is given in this issue.

**Comment by Huang-Wei** [comment]
I read it differently. I wrote an example https://play.golang.org/p/qns6XeQGcOE, it shows deepest trace at L12, rather than L17.

**Comment by Huang-Wei** [comment]
/reopen

I will raise cherrypick PRs to 1.12 and 1.11, and close this issue by then.

**Comment by Deepak1100** [comment]
Hi 
i have scaled my cluster to 100 node from 40 nodes as there were 400 pending pods and for some of the nodes all pods get schedule properly while around 28 nodes have only dnsmasq and kube-proxy was running.
after around 40 min those node had weave and other pods started running.
i coundn't able to find any other issue in my cluster apart from kube-scheduler got restarted because same error mentino in the issue.
though it restarted in few second.
```
E0111 10:18:37.611974       1 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/workspace/anago-v1.12.1-beta.0.52+4ed3216f3ec431/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/workspace/anago-v1.12.1-beta.0.52+4ed3216f3ec431/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/workspace/anago-v1.12.1-beta.0.52+4ed3216f3ec431/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:573
/usr/local/go/src/runtime/panic.go:502
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:388
/workspace/anago-v1.12.1-beta.0.52+4ed3216f3ec431/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/cache/node_info.go:387
/workspace/anago-v1.12.1-beta.0.52+4ed3216f3ec431/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:966
/workspace/anago-v1.12.1-beta.0.52+4ed3216f3ec431/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:888
/workspace/anago-v1.12.1-beta.0.52+4ed3216f3ec431/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:63
/usr/local/go/src/runtime/asm_amd64.s:2361
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x130dc2a]

goroutine 58786698 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/workspace/anago-v1.12.1-beta.0.52+4ed3216f3ec431/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x107
panic(0x15ca620, 0x247ea40)
	/usr/local/go/src/runtime/panic.go:502 +0x229
k8s.io/kubernetes/pkg/scheduler/cache.(*NodeInfo).Clone(0x0, 0xc42333ac80)
	/workspace/anago-v1.12.1-beta.0.52+4ed3216f3ec431/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/cache/node_info.go:387 +0x3a
k8s.io/kubernetes/pkg/scheduler/core.selectVictimsOnNode(0xc423c9f180, 0x1959380, 0xc422b8ebd0, 0x0, 0xc420598030, 0x197cbc0, 0xc42014c1c0, 0xc421af9ec0, 0x6, 0x8, ...)
	/workspace/anago-v1.12.1-beta.0.52+4ed3216f3ec431/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:966 +0x7d
k8s.io/kubernetes/pkg/scheduler/core.selectNodesForPreemption.func1(0x0)
	/workspace/anago-v1.12.1-beta.0.52+4ed3216f3ec431/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:888 +0x176
k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.ParallelizeUntil.func1(0xc421d96fa0, 0xc4239a8000, 0xc423ad4718, 0xc42139cd80)
	/workspace/anago-v1.12.1-beta.0.52+4ed3216f3ec431/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:63 +0x91
created by k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.ParallelizeUntil
	/workspace/anago-v1.12.1-beta.0.52+4ed3216f3ec431/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:55 +0x152
```

i was just wondering that if after restart would it takes 30-40 min to schedule pods? cause after 40 min everything seems to be working properly.

**Comment by Huang-Wei** [comment]
Close this as fix has been applied to 1.11+ k8s versions.

/close
